### DISCUSSION TITLE
Linkwarden: Update > 2.8.2 needs Monolith

### DISCUSSION THREAD

**Discussion by MickLesk** [community-scripts/ProxmoxVE#305] [Guides]
Hello,

if you update Linkwarden to a newer Version as 2.8.2 you need "Monolith" as Dependency.

Here is an short guide to install monolith:

Before:
- Check your hdd storage for the LXC. Cargo needs some ressources. Best is, to increase to 12GB Storage

1. Install Cargo & make dependency
```bash
apt-get install -y \
  make \
  cargo
```

2. Install Monolith with cargo
```bash
cargo install monolith
echo 'export PATH=~/.cargo/bin:$PATH' >>~/.bashrc
export PATH=~/.cargo/bin:$PATH
```

3. Reboot LXC

**Comment by HollyFredD** [comment]
Thanks mate

**Comment by knario** [comment]
thank you very much for all your help. It has worked very well.
Have a wonderful week.

**Comment by Schinken6** [comment]
Thanks for the help.
Will this be included in the script at some point or does this always have to be done manually?

**Comment by MickLesk** [comment]
Only for new Installations include

**Reply by rrik** [reply]
This only needs to be done on new installations? Or, is monolith included on all new installations going forward?

**Reply by MickLesk** [reply]
New Install. Include this.

**Reply by republicandaddy** [reply]
I tried installing using the script, but monolith won't install because the script installs version 1.75.0 of rustc whereas 1.80.0 or higher is require.  I tried to modify the script to add the following:

msg_ok "Updating rustc"
$STD curl https://sh.rustup.rs -sSf | sh
msg_ok "Updated rustc"

line 40, just before it installs monolith, but that didn't work.  I'm not a scripting expert and it tried to install everything on my proxmox server instead of creating a new lxc.  I don't know what I did to cause that.  Can you give me some direction on inserting this correctly so I can get this up and running?

Forgot to add...This is the error I get when using the script to setup the lxc...

  Downloaded 158 crates (22.6 MB) in 1.24s (largest was `openssl-src` at 10.0 MB)
error: failed to compile `monolith v2.8.3`, intermediate artifacts can be found at `/tmp/cargo-installoeHlyV`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.

Caused by:
  package `native-tls v0.2.14` cannot be built because it requires rustc 1.80.0 or newer, while the currently active rustc version is 1.75.0
  Try re-running cargo install with `--locked`
 ⠦
[ERROR] in line 41: exit code 0: while executing command $STD cargo install monolith

based on my doing some searching, this leads me to see that the wrong version of rustc is being installed.

**Reply by kappafon** [reply]
@republicandaddy

Also did not have any luck with the current script for fresh install.
Ended up doing a manual install, wasn't so bad.

Until the script is fixed, leaving the steps here if you want to try them.

---

First I created a [PostgreSQL LXC](https://community-scripts.github.io/ProxmoxVE/scripts?id=postgresql), including adminer (used adminer to create a new empty `linkwarden` database)

Then an [Ubutnu LXC](https://community-scripts.github.io/ProxmoxVE/scripts?id=ubuntu) (for linkwarden manual install)
Used advanced settings during install and mirrored the Linkwarden script setup (Disk: 12, RAM: 2048, Core: 2)

### Commands to run within Ubuntu (Linkwarden) LXC

Add Yarn APT repository to your system:
```bash
apt-get install -y gnupg
curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -
echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list
```

Update your package list:
```bash
apt-get update
```

Install the required dependencies:
```bash
apt-get install -y \
git \
nodejs \
npm \
cargo \
postgresql-client
```

Now clone the repository into your container:
```bash
git clone https://github.com/linkwarden/linkwarden.git
cd linkwarden
```

Install Yarn:
```bash
apt-get install -y yarn
```

Install project dependencies with Yarn:
```bash
yarn
npx playwright install --with-deps chromium
yarn playwright install
```

Generate a secret for the next step:
```bash
node -e "console.log(require('crypto').randomBytes(32).toString('hex'))"
```

Configure the Environment Variables:
```bash
cp .env.sample .env
nano .env
```

```ini
NEXTAUTH_SECRET=<generated_secret>
NEXTAUTH_URL=http://<linkwarden_ip>:3000/api/v1/auth
DATABASE_URL=postgresql://postgres:<password>@<postgresql_ip>:5432/<database_name>
```

Verify linkwarden connection (optional):
```
PGPASSWORD=<password> psql -h <postgresql_ip> -U postgres -d <database_name>
```

```bash
\q
```

Build and setup the database:
```bash
cd /root/linkwarden
yarn build
yarn prisma migrate deploy
yarn prisma generate
```

Fire it up:
```bash
yarn start
```

---
## Set Up the Systemd Service to trigger yarn start

```bash
nano /etc/systemd/system/linkwarden.service
```

```ini
[Unit]
Description=Linkwarden App
After=network.target

[Service]
WorkingDirectory=/root/linkwarden
ExecStart=/usr/bin/yarn start
Restart=always
User=root
Group=root
Environment=PATH=/usr/bin:/usr/local/bin
Environment=NODE_ENV=production

[Install]
WantedBy=multi-user.target
```

Reload systemd to Recognize the New Service:
```bash
systemctl daemon-reload
```

Enable the Service to Start on Boot:
```bash
systemctl enable linkwarden.service
```

Reboot the container.

**Comment by arl8** [comment]
Getting the below error, any help is appreciated! TIA!

  ✔️  Stopped Linkwarden
 ⠋mv: cannot stat '/opt/linkwarden/.env': No such file or directory

[ERROR] in line 42: exit code 0: while executing command mv /opt/linkwarden/.env /opt/.env

**Comment by BBT-nZ** [comment]
> Hello,
> 
> if you update Linkwarden to a newer Version as 2.8.2 you need "Monolith" as Dependency.
> 
> Here is an short guide to install monolith:
> 
> Before:
> 
> * Check your hdd storage for the LXC. Cargo needs some ressources. Best is, to increase to 12GB Storage
> 
> 1. Install Cargo & make dependency
> 
> ```shell
> apt-get install -y \
>   make \
>   cargo
> ```
> 
> 2. Install Monolith with cargo
> 
> ```shell
> cargo install monolith
> echo 'export PATH=~/.cargo/bin:$PATH' >>~/.bashrc
> export PATH=~/.cargo/bin:$PATH
> ```
> 
> 3. Reboot LXC

Hey! Still facing issues while trying to get the script to run. 

Are you able to point out in which step to take first? 

For e.g. Run the script in Proxmox Shell, towards the end of LXC creation it would error out. What next from here? 

Would appreciate a hand. Thankyou

RIP @tteck The Master. We miss you.

**Reply by MickLesk** [reply]
and whats the problem? This need to be executed first. In the Shell himself. Not with pct enter

**Reply by BBT-nZ** [reply]
I followed the steps to the letter.

1) Ran your commands first in Proxmox Shell. Once completed checked the rust version.
2) Ran the LXC script creation command under Proxmox Shell.

It fails at this step.

` ⠹  Downloaded 158 crates (22.6 MB) in 1.06s (largest was `openssl-src` at 10.0 MB)
error: failed to compile `monolith v2.8.3`, intermediate artifacts can be found at `/tmp/cargo-installDPmbpK`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.

Caused by:
  package `native-tls v0.2.14` cannot be built because it requires rustc 1.80.0 or newer, while the currently active rustc version is 1.75.0
  Try re-running cargo install with `--locked`
 ⠸
[ERROR] in line 41: exit code 0: while executing command $STD cargo install monolith
root@NB-PXMX-P340:~# ```

**Reply by MickLesk** [reply]
Before the other:

curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh

source $HOME/.cargo/env

**Reply by BBT-nZ** [reply]
I just removed everything and ran the prompts in the order below under Proxmox Shell:


1. `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs/ | sh`

`source $HOME/.cargo/env`

2. `apt-get install -y \
  make \
  cargo`

3. `cargo install monolith
echo 'export PATH=~/.cargo/bin:$PATH' >>~/.bashrc
export PATH=~/.cargo/bin:$PATH`

4. Reboot LXC


**Back at this error below:**

_⠇  Downloaded 158 crates (22.6 MB) in 1.07s (largest was `openssl-src` at 10.0 MB)
 ⠏error: failed to compile `monolith v2.8.3`, intermediate artifacts can be found at `/tmp/cargo-installnz9cDW`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.

Caused by:
  package `native-tls v0.2.14` cannot be built because it requires rustc 1.80.0 or newer, while the currently active rustc version is 1.75.0
  Try re-running cargo install with `--locked`
 ⠋
[ERROR] in line 41: exit code 0: while executing command $STD cargo install monolith_

**Reply by MickLesk** [reply]
rustup update

rustc --version

(its 1.80?)

rustup install 1.80.0
rustup default 1.80.0

**Reply by BBT-nZ** [reply]
Just wanted to hop in provide an update. I've now completely remove rust and cargo from my Proxmox Shell to start fresh. Before i do so, are you able to share in numerical order what commands to execute? Thanks a bunch for all your help so far.

**Comment by bobmarley2021** [comment]
Any update on this? :)

I can't login to run any of the commands because the script didn't finish successfully. No autologin and no idea what the root pass is. Unless anyone here has an idea?

**Reply by MickLesk** [reply]
Then is your Install broken, use an Backup and ignore the Update

**Reply by kappafon** [reply]
It might not be what you are looking for, but I summarized manual steps of setting up Linkwarden as an LXC
https://github.com/community-scripts/ProxmoxVE/discussions/305#discussioncomment-12285517

As a side note. I noticed a few other community scripts don't have autologin working after setup, so I'd find myself removing the container re-running the script with advanced settings so that I can specify a root password.

In the end I still find the scripts useful as a reference guide, but started setting up my containers from scratch. Many of the scripts are opinionated and will install additional things required to run them, things that aren't always optional. But you might already have setup as a separate LXC, for example postgreSQL, caddy...

**Reply by MickLesk** [reply]
All LXCs has Autologin in proxmox webgui. If Not. The Install is broken and need to be fixed.

**Reply by bobmarley2021** [reply]
@MickLesk This is a fresh install.. no backup. No possibility of changing root password.

@kappafon I have re-run the script 5 times, it never has a different outcome. It errors on Monolith and I cannot login to the container to troubleshoot due to not knowing root password. Thanks for the manual pointers, but I would not know how to setup the new empty linkwarden db for example.

I'll just wait for the script to be fixed.

**Reply by MickLesk** [reply]
How you execute the Shell where Login prompt shown?

**Reply by kappafon** [reply]
@bobmarley2021 
Got it. It can be tricky sometimes, and it sucks when things break. I'm sure the script will get fixed soon.
But if you feel up to it, here are the steps for setting up the database.

Regarding the root password. You can always run the scripts using **Advanced Settings**:
![firefox_ciRWgG01kI](https://github.com/user-attachments/assets/9979ec5f-faeb-487d-a73f-f8f57f90aa90)

Then just keep pressing enter, with the exception of setting the password, here you could define it.
All the other script defaults are always preselected. If it's empty just leave it empty.

For the database I actually used the [postgresql script](https://community-scripts.github.io/ProxmoxVE/scripts?id=postgresql), including `Adminer` web UI (offered during script installation).

There are only a few steps afterwards.

Set a password after installation for postgres user by running:

```bash
echo "ALTER USER postgres with encrypted password 'somePassword';" | sudo -u postgres psql
```
And visit the Adminer in the browser http://\<container-ip\>/adminer

Fill it out:
![firefox_cVvkXR9y4F](https://github.com/user-attachments/assets/48cbcc74-1aeb-4b0d-beae-f98816531efa)

Again, Server is your container IP, Password is the one you setup by running the above command, and Database input can be empty.

Then you just click **Create database** and give it a name, e.g. _linkwarden_

That should be it.

With that you define the env variable in linkwarden container
```ini
DATABASE_URL=postgresql://postgres:<somePassword>@<postgresql_ip>:5432/<database_name>
```
so it would look something like this:
```ini
DATABASE_URL=postgresql://postgres:somePassword@192.168.1.11:5432/linkwarden
```

**Comment by bobmarley2021** [comment]
Thank you both for the prompt replies again. 

@kappafon That's amazing, thank you. I really appreciate the step by step :+1: 

@MickLesk to execute the script ( `bash -c "$(wget -qLO - https://github.com/community-scripts/ProxmoxVE/raw/main/ct/linkwarden.sh)"` ), I click on Node > Shell, then paste the command, when the error occurs, the newly created LXC container is shown in the list under the node but when clicking on it, the login prompt is shown which I cannot pass. Is this what you meant?

Edit: to get to the LXC shell, I just click on [LXC name] > Console
